extension/src/time_weighted_average.rs

```diff
@@ -0,0 +1,422 @@
+            let state = match state {
+                None => return None,
+                Some(state) => state,
+            };
+
+            flatten!(
+                time_weight_summary {
+                    method: &state.method,
+                    first: &state.first,
+                    last: &state.last,
+                    w_sum: &state.w_sum,
+                }
+            ).into()
+        })
+    }
+}
+
+extension_sql!(r#"
+CREATE OR REPLACE FUNCTION time_weight_summary_in(cstring) RETURNS time_weight_summary IMMUTABLE STRICT PARALLEL SAFE LANGUAGE C AS 'MODULE_PATHNAME', 'time_weight_summary_in_wrapper';
+CREATE OR REPLACE FUNCTION time_weight_summary_out(time_weight_summary) RETURNS CString IMMUTABLE STRICT PARALLEL SAFE LANGUAGE C AS 'MODULE_PATHNAME', 'time_weight_summary_out_wrapper';
+
+CREATE TYPE time_weight_summary (
+    INTERNALLENGTH = variable,
+    INPUT = time_weight_summary_in,
+    OUTPUT = time_weight_summary_out,
+    STORAGE = extended
+);
+
+CREATE AGGREGATE time_weight_ordered(method text, ts timestamptz, value DOUBLE PRECISION)
+(
+    sfunc = time_weight_ordered_trans,
+    stype = internal,
+    finalfunc = time_weight_ordered_final,
+    combinefunc = time_weight_ordered_combine,
+    serialfunc = time_weight_summary_serialize,
+    deserialfunc = time_weight_summary_deserialize,
+    parallel = restricted
+);
+
+CREATE AGGREGATE time_weight_ordered(tws time_weight_summary)
+(
+    sfunc = time_weight_summary_ordered_trans,
+    stype = internal,
+    finalfunc = time_weight_ordered_final,
+    combinefunc = time_weight_ordered_combine,
+    serialfunc = time_weight_summary_serialize,
+    deserialfunc = time_weight_summary_deserialize,
+    parallel = restricted
+);
+"#);
+
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct TimeWeightTransState {
+    #[serde(skip)]
+    buffer: Vec<TSPoint>,
+    method: Option<TimeWeightMethod>,
+    combine_buffer: Vec<TimeWeightSummary>,
+}
+
+impl TimeWeightTransState {
+    fn push(&mut self, value: TSPoint, method: TimeWeightMethod) {
+        match self.method {
+            None => {self.method = Some(method)},
+            Some(m)=> {if m != method {panic!("Mismatched methods")}}
+        }
+        self.buffer.push(value);
+    }
+
+    fn sort_and_calc(&mut self) {
+        if self.buffer.is_empty() {
+            return
+        }
+        self.buffer.sort_unstable_by_key(|p| p.ts);
+        match self.method {
+            None => panic!("invalid state"), // this shouldn't be None if the buffer is not empty
+            Some(m)=>self.combine_buffer.push(TimeWeightSummary::new_from_sorted_iter(&self.buffer, m).unwrap()),
+        };
+        self.buffer.clear();
+    }
+    fn combine_push(&mut self, other: &TimeWeightTransState) {
+        let cb = other.combine_buffer.clone();
+        for val in cb.into_iter(){
+            self.combine_buffer.push(val);
+        };
+    }
+    fn combine_sort_and_calc(&mut self) {
```

Comment:
Alright, these function names are painful, and it might just come from how your variables are named.  Maybe `computed_summaries` or `summary_buffer` instead of `combine_buffer` and then `push_summary` and `combine_summaries` for the functions affecting it.

Reply:
modified the names.